Add protobuf for capture

IrinaShkviro opened the pull request:
This PR replace cereal with protobuf in capture serialization. FunctionStats, Function, CallstackEvent and LinuxAddressInfo were replaced with proto messages. Added Callstack, CaptureHeader and TimerData proto messages. All new protos are in the OrbitClientProtos.

Also, save the version with the captures to have better error messages for user when loading incompatible captures. 

The capture consists of CaptureHeader, CaptureData and Timers. Capture data are copied to protobuf message for now.

IrinaShkviro commented on protos/timer.proto:
```diff
@@ -0,0 +1,27 @@
+
+syntax = "proto3";
+
+message TimerData {
+  uint64 start = 1;
+  uint64 end = 2;
+  int32 process_id = 3;
+  int32 thread_id = 4;
+  uint32 depth = 5;
+
+  enum Type {
+    kNone = 0;
+    kCoreActivity = 1;
+    kIntrospection = 2;
+    kGpuActivity = 3;
+  }
+  Type type = 6;
+
+  int32 processor = 7;
+  uint64 callstack_hash = 8;
+  uint64 function_address = 9;
+  uint64 user_data_key = 10;
+  uint64 timeline_hash = 11;
+}
```
@dpallotti I see your comment about completely changing the TimerData in the previous PR, but I would prefer to postpone it for now, because it requires a lot of changes

beckerhe commented on protos/function.proto:
```diff
@@ -11,3 +11,33 @@ message FunctionStats {
+	kOrbitTrackFloat = 9;
+	kOrbitTrackDouble = 10;
+	kOrbitTrackFloatAsInt = 11;
+	kOrbitTrackDoubleAsInt64 = 12;
```
I think clang-format does only support languages supported by clang, so no proto. But we should follow https://developers.google.com/protocol-buffers/docs/style?hl=en

antonrohr replied:
$ clang-format -help
OVERVIEW: A tool to format C/C++/Java/JavaScript/Objective-C/__Protobuf__/C# code.

antonrohr replied:
My vscode "Format Document" (via clangd extension) works out of the box. I dont know about VS or CLion

dpallotti replied:
Ctrl+Alt+L in CLion works fine for me with protos.

IrinaShkviro commented on protos/function.proto:
```diff
@@ -0,0 +1,43 @@
+
+message FunctionStats {
+  uint64 count = 1;
+  double total_time_ms = 2;
+  double average_time_ms = 3;
+  double min_ms = 4;
+  double max_ms = 5;
```
@dpallotti suggested to use uint64 and nanosecond for these fields

IrinaShkviro replied:
Updated these fields

antonrohr commented on OrbitCore/SamplingProfiler.cpp:
```diff
@@ -217,13 +218,13 @@ void SamplingProfiler::ProcessSamples() {
 
-    for (uint32_t i = 0; i < callstack->m_Data.size(); ++i) {
-      uint64_t addr = callstack->m_Data[i];
+    for (int i = 0; i < callstack->data_size(); ++i) {
```
Maybe
```suggestion
    for (uint64_t address: callstack->mutable_data()) 
```

antonrohr commented on OrbitCore/Capture.cpp:
```diff
@@ -237,14 +239,14 @@ void Capture::RegisterZoneName(uint64_t a_ID, const char* a_Name) {
 
 //-----------------------------------------------------------------------------
-void Capture::AddCallstack(CallStack& a_CallStack) {
+void Capture::AddCallstack(HashedCallstack& a_CallStack) {
```
```suggestion
void Capture::AddCallstack(HashedCallstack& callstack) {
```

IrinaShkviro replied:
I think these classes require some additional polishing. I reverted most of the changes for now as those changes are not necessary for this task. So would you mind if I do it in the separate PR?

antonrohr commented on OrbitCore/Capture.cpp:
```diff
@@ -237,14 +239,14 @@ void Capture::RegisterZoneName(uint64_t a_ID, const char* a_Name) {
 
 //-----------------------------------------------------------------------------
-std::shared_ptr<CallStack> Capture::GetCallstack(CallstackID a_ID) {
+std::shared_ptr<HashedCallstack> Capture::GetCallstack(CallstackID a_ID) {
```
```suggestion
std::shared_ptr<HashedCallstack> Capture::GetCallstack(CallstackID hash) {
```
or callstack_hash

antonrohr commented on OrbitCore/Capture.h:
```diff
@@ -63,13 +63,14 @@ class Capture {
   static std::shared_ptr<Process> GTargetProcess;
   static std::shared_ptr<Preset> GSessionPresets;
-  static std::shared_ptr<CallStack> GSelectedCallstack;
+  static std::shared_ptr<HashedCallstack> GSelectedCallstack;
```
I think this global is only used in CaptureWindow and there just in two lines right next to each other. I am in favor of removing it.

IrinaShkviro replied:
Yes, and the same for Capture::GCallstacks; probably, I should remove both.

antonrohr commented on OrbitCore/SamplingProfiler.cpp:
```diff
@@ -177,16 +178,16 @@ void SamplingProfiler::ProcessSamples() {
       std::set<uint64_t> uniqueAddresses;
-      for (uint32_t i = 0; i < resolvedCallstack->m_Data.size(); ++i) {
-        uniqueAddresses.insert(resolvedCallstack->m_Data[i]);
+      for (int i = 0; i < resolvedCallstack->data_size(); ++i) {
+        uniqueAddresses.insert(resolvedCallstack->data(i));
```
insert a copy of the whole vector instead of each element

IrinaShkviro commented:
> I somehow started going through this like a regular PR. Lets talk about this in the meeting and then I continue

Yes, let's discuss it in the meeting, then I'll update PR and mark it as "ready for review".

IrinaShkviro commented on protos/capture_data.proto:
```diff
@@ -13,3 +13,33 @@ message FunctionStats {
   uint64 max_ns = 5;
 }
+
+message FunctionInfo {
```
FunctionInfo was introduced only to save hooked functions (that are stored in GSelectedFunctionsMap). Fields `file`, `line`, `type` and `pretty_name` aren't used in context of hooked functions, but the `stats` field is only useful for hooked functions (in other cases it's not filled).

IrinaShkviro commented on protos/capture_data.proto:
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+message CaptureHeader {
+  string capture_name = 1;
```
Yes, it's just a filename and used only in Save(). As we don't use it when loading capture we can remove it for now.

IrinaShkviro commented on protos/callstack.proto:
```diff
@@ -9,3 +9,8 @@ message CallstackEvent {
 }
+
+message HashedCallstack {
+  uint64 callstack_hash = 1;
```
`callstack_hash` is calculated based on data vector, so storing only the vector would be enough. Updated the PR and replaced this proto message with `CallstackInfo`

IrinaShkviro commented on protos/address_info.proto:
```diff
@@ -0,0 +1,12 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Added all of them into one new file, but from the other hand, it's a bit confusing to include "capture_data.pb.h" when you need just CallstackEvent or FunctionInfo, for example.

IrinaShkviro commented:
The changes are:
- Moved all the protos in the one file (`capture_data.proto`)
- Renamed proto messages
- Introduced CallstackInfo and save data without hash instead of replacing current CallStack struct (with `hash` field)
- Used nanoseconds instead of milliseconds in FunctionStats
- didn't change Function and FunctionsStats; only translated these structures to the proto messages
- added a package specifier in the new proto file

dpallotti commented on protos/capture_data.proto:
```diff
@@ -0,0 +1,15 @@
+
+syntax = "proto3";
+
+package protos;
```
Don't we want a more specific namespace name that refers to the fact that these are the protos for file serialization? Otherwise all protos should end up in this same namespace and I assume that's not what we want?

IrinaShkviro replied:
Yes, and also I should move `capture_data` to another folder. But I'm still thinking what name/structure would be better: `protos.ui.serialization` / `protos.file_serialization` / `protos.serialization.file`?

dpallotti commented on OrbitCore/Utils.h:
```diff
@@ -290,6 +291,10 @@ inline std::string GetPrettySize(uint64_t size) {
   return absl::StrFormat("%.2f TB", size / TB);
 }
 
+inline double GetMillisecondsFromNanoseconds(uint64_t nanoseconds) {
```
Do we really want this method to wrap an `absl` one-liner into another one-liner?

dimitry- replied:
if you decide to leave it probably rename to NanosecondsToMilliseconds()

IrinaShkviro replied:
Let's just remove it for now. Ideally, I wanted to have time always in one format (nanoseconds for example) and then make a conversion only when show time to the user. In this case, we can localize all the conversion in one place.

dpallotti commented on OrbitCore/FunctionUtils.cpp:
```diff
@@ -14,95 +14,114 @@
 }
 
-void Select(Function* func) {
+std::shared_ptr<protos::FunctionInfo> CreateFunction(
```
`CreateFunctionInfo`?

IrinaShkviro replied:
Yes, thanks, fixed it

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -72,15 +72,15 @@ void CaptureSerializer::SaveImpl(T& archive) {
     }
 
-    archive(functions);
+    // archive(functions);
```
Why is this commented?

IrinaShkviro replied:
I removed all this stuff in c431424dc4d4d772a8585c33631ceb205c8eb9c7 when added serialization with protobuf, but it's still a part of the history. I'll update commits after the review.

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -26,87 +30,107 @@
-  cereal::BinaryOutputArchive archive(stream);
   std::basic_ostream<char> Stream(&GStreamCounter);
-  cereal::BinaryOutputArchive CountingArchive(Stream);
+  google::protobuf::io::OstreamOutputStream out_counting_stream(&Stream);
+  google::protobuf::io::CodedOutputStream coded_counting_output(
+      &out_counting_stream);
+
   GStreamCounter.Reset();
-  SaveImpl(CountingArchive);
+  SaveImpl(&coded_counting_output);
   PRINT_VAR(GStreamCounter.Size());
   GStreamCounter.Reset();
```
I know it has always been there, but honestly do we even need this part? This serializes just to print the size? It's probably quite an unnecessary overhead.

IrinaShkviro replied:
I don't know if somebody uses it. But it looks like I already removed the useful part and now it shows only the total  size which is not so interesting. Btw, if we need the size we always can just change WriteMessage and log message_size.

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -138,12 +159,24 @@ ErrorMessageOr<void> CaptureSerializer::Load(const std::string& filename) {
+    return false;
   }
+
+  char* buffer = new char[message_size];
```
Is this a case for `make_unique_for_overwrite` that is currently in `OrbitLinuxTracing`? Maybe we should move it to `OrbitBase`?

dimitry- replied:
yes - it also provides scoping of the buffer

IrinaShkviro replied:
It's done, thanks!

dimitry- commented on protos/capture_data.proto:
```diff
@@ -0,0 +1,15 @@
+  uint64 average_time_ns = 3;
+  uint64 min_ns = 4;
+  uint64 max_ns = 5;
+}
```
maybe move these to OrbitGl? they are not used on the service side are they?

IrinaShkviro replied:
These messages aren't used on the service side. Does `protos` only for service? I thought that it's for all the protos, and because of it placed `capture_data` here. But talking about moving it to another module: why to OrbitGl and not to OrbitCore? We use CallstackEvent and FunctionInfo in OrbitCore, so probably we should move to OrbitCore or change the logic.

dimitry- commented on OrbitCore/Profiling.h:
```diff
@@ -28,6 +28,10 @@ inline TickType OrbitTicks(uint32_t a_Clock = 1 /*CLOCK_MONOTONIC*/) {
   return 1000000000ll * ts.tv_sec + ts.tv_nsec;
 }
 
+inline uint64_t NanoSecondsFromTicks(TickType start, TickType end) {
```
TicksToNanoseconds? or at least s/NanoSeconds/Nanoseconds/

antonrohr replied:
Since the other methods in this class are also using the from naming, I would be okay with leaving it as is. Or changing all the names (maybe todo message -> other PR)

IrinaShkviro replied:
I changed the name to `TicksToNanoseconds` and also added TODO in the code.

dimitry- commented on protos/capture_data.proto:
```diff
@@ -13,3 +13,33 @@ message FunctionStats {
+  }
+  OrbitType type = 10;
+  FunctionStats stats = 11;
+}
```
looking at existing model - seems like Function should be an aggregation of SymbolInfo, ModuleSymbols & FunctionStats, or am I missing something?

IrinaShkviro replied:
let's discuss it offline with @antonrohr

dimitry- commented on OrbitGl/LiveFunctionsDataView.cpp:
```diff
@@ -327,14 +326,15 @@ void LiveFunctionsDataView::OnTimer() {
 
 //-----------------------------------------------------------------------------
-Function& LiveFunctionsDataView::GetFunction(unsigned int a_Row) const {
+protos::FunctionInfo& LiveFunctionsDataView::GetFunction(
```
should probably be a pointer.

IrinaShkviro replied:
Yes, thanks!

dimitry- commented on OrbitGl/LiveFunctionsDataView.cpp:
```diff
@@ -327,14 +326,15 @@ void LiveFunctionsDataView::OnTimer() {
 
 std::pair<TextBox*, TextBox*> LiveFunctionsDataView::GetMinMax(
-    Function& function) const {
+    protos::FunctionInfo& function) const {
```
const?

IrinaShkviro replied:
Fixed it

dimitry- commented on protos/capture_data.proto:
```diff
@@ -56,3 +56,7 @@ message LinuxAddressInfo {
 }
+
+message CaptureHeader {
+  uint32 version = 1;
```
add/replace with string, we need something to show to users. (This capture was saved with v1.42 earliest supported version is v1.80...)

dimitry- commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -26,87 +30,107 @@
 #include "capture_data.pb.h"
 
+static absl::flat_hash_map<const uint32_t, const std::string>
+    capture_version_to_orbit_version_map{
```
this also works.. but now we need to keep int -> string mapping in all of our versions, where if we save it to proto - we will only need to keep the latest supported one in the code.

IrinaShkviro replied:
One more idea: do we want to show the exact version for the user, or the range? For example, "this capture could be opened with v1.42-v1.45" (for this I should change the type of the map and the error message). In this case if capture was taken in 1.43, we won't require v1.43 to open the capture, but any version from the range. But we don't know the max version when we save the capture, and for this purpose mapping could be useful.

IrinaShkviro replied:
I added the solution with the range of the appropriate Orbit versions that gives more information to the user.

dimitry- commented on OrbitCore/OrbitModule.cpp:
```diff
@@ -53,7 +53,7 @@ void Module::LoadSymbols(const ModuleSymbols& module_symbols) {
+        FunctionUtils::CreateFunctionInfo(
         symbol_info.name(), symbol_info.demangled_name(), symbol_info.address(),
         module_symbols.load_bias(), symbol_info.size(),
         symbol_info.source_file(), symbol_info.source_line(), m_FullName,
```
m_FullName and m_AddressStart should also be used from symbol_info.

dimitry- replied:
which also means this function is not using anything from Module class.

antonrohr commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -24,114 +28,94 @@
-    ORBIT_SIZE_SCOPE("Capture::GProcessId");
-    archive(Capture::GProcessId);
+  capture_info->mutable_address_infos()->Reserve(Capture::GAddressInfos.size());
+  for (const auto& address_indo : Capture::GAddressInfos) {
```
```suggestion
  for (const auto& address_info : Capture::GAddressInfos) {
```

IrinaShkviro replied:
Fixed, thanks!

antonrohr commented on OrbitGl/LiveFunctionsDataView.cpp:
```diff
@@ -49,28 +51,33 @@ std::string LiveFunctionsDataView::GetValue(int a_Row, int a_Column) {
     case COLUMN_TIME_TOTAL:
-      return GetPrettyTime(stats->m_TotalTimeMs);
+      return GetPrettyTime(
+          absl::ToDoubleMilliseconds(absl::Nanoseconds(stats.total_time_ns())));
     case COLUMN_TIME_AVG:
-      return GetPrettyTime(stats->m_AverageTimeMs);
+      return GetPrettyTime(absl::ToDoubleMilliseconds(
+          absl::Nanoseconds(stats.average_time_ns())));
     case COLUMN_TIME_MIN:
-      return GetPrettyTime(stats->m_MinMs);
+      return GetPrettyTime(
+          absl::ToDoubleMilliseconds(absl::Nanoseconds(stats.min_ns())));
     case COLUMN_TIME_MAX:
-      return GetPrettyTime(stats->m_MaxMs);
+      return GetPrettyTime(
+          absl::ToDoubleMilliseconds(absl::Nanoseconds(stats.max_ns())));
```
I am aware of @dpallotti comment from earlier. This here seems like an unnecessary amount of code duplication. I would propose to add a second GetPrettyTime to Utils.h
`std::string GetPrettyTime(uint64_t nanoseconds) { ... }`

IrinaShkviro replied:
Ideally, I wanted to replace `std::string GetPrettyTime(double milliseconds)` with `std::string GetPrettyTime(uint64_t nanoseconds)` in the next PR to reduce amount  of changes. But probably I can do it in two steps (just add a second function for now and then get rid of the first one).

beckerhe replied:
Please no overloads for two numeric types with different units. It's bad enough to convey the unit in the name of the parameter. GetPrettyTime should take a absl::Duration.

IrinaShkviro replied:
I like the idea to use absl::Duration, and it doesn't require too much changes. I added a commit with this change (469cf9900af0df4f43f87b3dbcbd7ad52765f700).

IrinaShkviro commented on OrbitCore/Utils.h:
```diff
@@ -31,6 +31,7 @@
 #include "absl/strings/match.h"
 #include "absl/strings/str_format.h"
 #include "absl/strings/str_split.h"
+#include "absl/time/time.h"
```
I should remove it in 7803da53851fecc6423121dc5445bcf10ca5a3c2, thanks for noticing!

IrinaShkviro commented on OrbitCore/SamplingProfiler.h:
```diff
@@ -74,8 +71,8 @@ class SamplingProfiler {
-  void AddCallStack(CallstackEvent& callstack_event);
-  void AddHashedCallStack(CallstackEvent& a_CallStack);
+  void AddCallStack(protos::CallstackEvent& callstack_event);
+  void AddHashedCallStack(protos::CallstackEvent& a_CallStack);
```
Thanks, renamed the argument.

dpallotti commented on OrbitUIProtos/capture_data.proto:
```diff
@@ -4,7 +4,7 @@
 syntax = "proto3";
 
-package protos;
+package orbit_ui_protos;
```
Soo, maybe this was discussed offline, but I'm not a big fun of this namespace name. "orbit" feels redundant and "ui" not too appropriate. Maybe "client_protos"?
Same for the module/directory name: I guess we want to keep "Orbit" for consistency with the other modules there, but maybe "OrbitClientProtos"?

IrinaShkviro replied:
I don't remember if we agree on the certain name, so we can discuss it now. @dimitry- @antonrohr @beckerhe any other suggestions?

dimitry- replied:
we haven't we just discussed separating it from current protos - client sounds good to me, but I do not have strong feelings one way or the other.

IrinaShkviro replied:
Ok, I'm not against OrbitClientProtos, and as no one objects, I renamed it to OrbitClientProtos.

dpallotti commented on OrbitUIProtos/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+
+cmake_minimum_required(VERSION 3.15)
+
+project(OrbitUIProtos)
```
As stated in the other comment, maybe "OrbitClientProtos"?

IrinaShkviro replied:
Renamed it, thanks!

dpallotti commented on OrbitCore/Profiling.h:
```diff
@@ -30,7 +30,8 @@ inline TickType OrbitTicks(uint32_t a_Clock = 1 /*CLOCK_MONOTONIC*/) {
 }
 
-inline uint64_t NanoSecondsFromTicks(TickType start, TickType end) {
+// TODO: use similar naming for other functions
```
Maybe it can just be done? It's only two functions.

IrinaShkviro replied:
It's done

dpallotti commented on OrbitCore/SamplingProfiler.cpp:
```diff
@@ -74,12 +74,12 @@ void SamplingProfiler::AddCallStack(CallstackEvent& callstack_event) {
 //-----------------------------------------------------------------------------
-void SamplingProfiler::AddHashedCallStack(CallstackEvent& a_CallStack) {
-  if (!HasCallStack(a_CallStack.callstack_hash())) {
+void SamplingProfiler::AddHashedCallStack(CallstackEvent& callstack_event) {
```
Are both `AddCallstack` and `AddHashedCallstack` needed now? After recent changes they conceptually do the same thing, except that the former acts when `!HasCallStack` and the latter prints an error. Can this one just be removed or am I missing something?

IrinaShkviro replied:
Removed AddHashedCallstack.

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -24,94 +29,101 @@
+
+static absl::flat_hash_map<const uint32_t,
+                           std::pair<const std::string, const std::string>>
+    capture_version_to_orbit_versions_map{
```
Maintaining this mapping and the ranges feels like quite a burden to me. Honestly I would lean towards Dimitry's proposal of just saving the version string in the protos.

IrinaShkviro replied:
It looks a bit confusing for now, because the minimum and maximum version are the same. I think about the situation when we have range of 10 for example captures from different versions that are compatible and trying to open it with a newer version of Orbit. For all 10 we will get the error saying the exact version and to open it we should have all the 10 versions (according to error). But if we show the range, the user knows that it's enough to have just one version for all the captures. But it depends also on how this map looks like: if we always have min = max it doesn't make sense at all.

dpallotti replied:
It's not so much about how it looks, I'm more unsure about whether it's useful to maintain such a map in relation to the fact that it has to be manually maintained and updated for every release. Considering that we're still in beta, we're not committing to maintaining compatibility yet, and that we are probably breaking compatibility at every release, maybe for user it's just better to get a message saying "This capture was created with version ... and is not compatible with the current version" (by saving the version string in the proto) than suggesting them a range of versions to use to open it.
But I might very well be wrong here. Does anyone else have any thoughts about this?

beckerhe replied:
True, having such a message might be very convenient for the user. So probably we want to introduce this map (or something similar at some point). But for me it feels a bit premature to do it now. We are not loosing anything by postponing the introduction of such a map to the point where we introduce the first (backwards-compatible) breaking change. And I hope at that point we will have a concept on how to handle breaking changes and for how long and so on.

IrinaShkviro replied:
Ok, let's postpone it for now. I removed the map and changed int to string for `version` field in "capture_data.proto".

IrinaShkviro commented on OrbitCore/Profiling.h:
```diff
@@ -37,7 +39,14 @@ inline double MicroSecondsFromTicks(TickType a_Start, TickType a_End) {
   return double((a_End - a_Start)) * 0.001;
 }
 
+inline double MilliSecondsFromTicks(TickType start, TickType end) {
```
Yes, I renamed functions in Profiling.h, thanks!

beckerhe commented on OrbitGl/CaptureWindow.cpp:
```diff
@@ -551,12 +554,12 @@ void CaptureWindow::Draw() {
     float to = time_graph_.GetWorldFromTick(maxTime);
 
-    double micros = MicroSecondsFromTicks(minTime, maxTime);
+    uint64_t nanos = TicksToNanoseconds(minTime, maxTime);
```
Maybe TicksToDuration? :innocent:

IrinaShkviro replied:
Done, thanks!

IrinaShkviro commented on OrbitCore/StringManager.h:
```diff
@@ -21,7 +20,9 @@ class StringManager {
   void Clear();
 
-  ORBIT_SERIALIZABLE;
+  const absl::flat_hash_map<uint64_t, std::string> GetKeyToStringMap() const {
```
Yes, thanks, fixed

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -190,31 +196,76 @@ ErrorMessageOr<void> CaptureSerializer::Load(std::istream& stream) {
+  if (header.version() != kRequiredCaptureVersion) {
+    std::string incompatible_version_error_message = absl::StrFormat(
+        "This capture format is no longer supported but could be opened with "
+        "%s Orbit version.",
```
I would maybe put the version at the end, e.g. "This capture format is no longer supported but could be opened with Orbit version %s."

IrinaShkviro replied:
Done

dpallotti commented on OrbitGl/CaptureSerializer.cpp:
```diff
@@ -190,31 +196,76 @@ ErrorMessageOr<void> CaptureSerializer::Load(std::istream& stream) {
+
+  std::string error_message =
+      "Error parsing the capture.\nNote: If the capture "
+      "was taken in a previous Orbit version, it could be incompatible. "
```
```suggestion
      "was taken with a previous Orbit version, it could be incompatible. "
```

IrinaShkviro replied:
Fixed, thanks!

IrinaShkviro commented:
Thanks a lot for the review! I rebased the branch and clean up commit history (probably too early, sorry for that 😬).
